lib/scholar/metrics/clustering.ex

```diff
@@ -0,0 +1,155 @@
+defmodule Scholar.Metrics.Clustering do
+  @moduledoc """
+  Metrics related to clustering algorithms.
+  """
+
+  import Nx.Defn
+
+  defnp inner_dist(x, labels) do
+    {num_samples, num_features} = Nx.shape(x)
+
+    expanded_points =
+      Nx.tile(x, [1, num_samples]) |> Nx.reshape({num_samples, num_samples, num_features})
+
+    labels_vectorize = Nx.broadcast(labels, {num_samples, num_samples}) |> Nx.transpose()
+
+    mask_zero_one =
+      labels_vectorize ==
+        Nx.tile(labels, [1, num_samples]) |> Nx.reshape({num_samples, num_samples})
+
+    denominator = Nx.sum(mask_zero_one, axes: [1]) |> Nx.reshape({num_samples, 1}) |> Nx.squeeze()
+    indices_to_zero = Nx.equal(denominator, 1)
+    denominator = Nx.select(Nx.greater(denominator, 1), denominator - 1, 1)
+
+    mask =
+      mask_zero_one
+      |> Nx.select(
+        Nx.tile(Nx.iota({num_samples}), [num_samples, 1]),
+        Nx.broadcast(Nx.iota({num_samples, 1}), {num_samples, num_samples})
+      )
+
+    points = Nx.take(x, mask)
+
+    {Nx.sum(
+       Scholar.Metrics.Distance.euclidean(
+         expanded_points,
+         points,
+         axes: [2]
+       ),
+       axes: [1]
+     ) / denominator, indices_to_zero}
+  end
```

Comment:
It seems to me that having the pairwise distance from `outer_dist` we can easily compute the `inner_dist` out of that. Perhaps something like this:

```elixir
inner_dist =
  dist_in_cluster
  |> Nx.divide(cluster_size - 1)
  |> Nx.take_along_axis(Nx.reshape(labels, {num_samples, 1}), axis: 1)
  |> Nx.squeeze(axes: [1])
```

`dist_in_cluster` stores total distance from datapoint to elements in the given cluster and we are only interested in the cluster that the datapoint is in (it's actually the one we set infinity for in `outer_dist`).